middle_end/flambda/types/env/typing_env_level.rec.ml

```diff
@@ -302,10 +304,27 @@ let join_types ~env_at_fork envs_with_levels ~extra_lifted_consts_in_use_envs =
       env_at_fork
       envs_with_levels
   in
+  (* Special handling for parameters: they're defined in [env_at_fork],
+     but their type (Unknown) is only a placeholder until we compute
+     the actual join.
```

Comment:
From this comment, I gather that there are two potential meanings for a value to have an `Unknown` type :
- the value can be anything / the type system imprecision cannot represent a more precise approximation (i.e. top)
- there is no information / no approximation for the given value (i.e. an absence of type)
And thus params are initially added with the second meaning, but during the join, `Unknown` types in general must be treated using the first meaning. Am I correct ?
If so, that would suggest we maybe need distinguished constructors to make the two meanings different types instead of overloading `Unknown`.

Reply:
Not really. The type `Unknown` has only one meaning: anything is possible.
However, in this particular case, we now that the type for the parameters must be refined by the join, so we start our join computation with Bottom types for the parameters.
The issue with the parameters is that they are added to `env_at_fork` quite early, so we get to manipulate an environment where they have `Unknown`type for a while before they can be properly computed by `join`.
They can't be treated exactly like variables defined in the branches either, as they need to be defined with mode `Normal` instead of `In_types` in the result environment.
Overall the design isn't perfect, but I think the main problem is the handling of the various kinds of variables, not the types themselves.

One other fix I considered was adding `Bottom` equations on the parameters when adding them to `env_at_fork`, but this (adding `Bottom` equations) is something I'd like to avoid in general (though one could argue that it's fine in this particular case, as the corresponding environment is indeed technically unreachable).